solutions/cfgym-101102D.mdx

```diff
@@ -0,0 +1,263 @@
+---
+id: cfgym-101102D
+source: CF
+title: Rectangles
+author: Kevin Sheng
+---
+
+Let's first simplify this problem, and then apply the solution of that problem to this one.
+
+Say we had a matrix of all `0`'s and `1`'s. How many submatrices of all `1`'s are there?
+
+We first define a 2D array, $\texttt{longest\_consec}$, that contains for each cell how many `1`'s there are to the right.  
+For example, this matrix:
+
+<center>
+<table>
+	<tr>
+		<td>1</td>
+		<td>0</td>
+		<td>1</td>
+		<td>1</td>
+	</tr>
+	<tr>
+		<td>0</td>
+		<td>1</td>
+		<td>1</td>
+		<td>1</td>
+	</tr>
+	<tr>
+		<td>0</td>
+		<td>0</td>
+		<td>0</td>
+		<td>1</td>
+	</tr>
+</table>
+</center>
+
+Would have the following $\texttt{longest\_consec}$:
+
+<center>
+<table>
+	<tr>
+		<td>1</td>
+		<td>0</td>
+		<td>2</td>
+		<td>1</td>
+	</tr>
+	<tr>
+		<td>0</td>
+		<td>3</td>
+		<td>2</td>
+		<td>1</td>
+	</tr>
+	<tr>
+		<td>0</td>
+		<td>0</td>
+		<td>0</td>
+		<td>1</td>
+	</tr>
+</table>
+</center>
+
+Now, we go through each column and then for each row in reverse,
+find out for each cell the number of all `1`'s submatrices where the cell is the upper left corner.
```

Comment:
this is worded a bit awkwardly

Reply:
lol it's the best i could do

Reply:
why does going through each column and then for each row in reverse mean

Reply:
so you iterate through each column
then for each column you go through each row in that column in reverse
the code explains it better tbh